src/lib/Incubator/Cache/TagAware/RedisTagAwareAdapter.php

```diff
@@ -0,0 +1,172 @@
+                $tagSets[$getId(self::TAGS_PREFIX . $tag)][] = $id;
+            }
+        }
+
+        // While pipeline isn't supported on RedisCluster, other setups will at least benefit from doing this in one op
+        $results = $this->pipeline(function () use ($serialized, $lifetime, $tagSets) {
+            // 1: Store cache items
+            foreach ($serialized as $id => $value) {
+                // Note: There is no MSETEX so we need to set each one
+                yield 'setEx' => [
+                    $id,
+                    0 >= $lifetime ? self::FORCED_ITEM_TTL : $lifetime,
+                    $value,
+                ];
+            }
+
+            // 2: append tag sets, method to add with array values differs on PHP Redis clients
+            $method = $this->redis instanceof Predis\Client ? 'sAdd' : 'sAddArray';
+            foreach ($tagSets as $tagId => $ids) {
+                yield $method => [$tagId, $ids];
+            }
+        });
+
+        foreach ($results as $id => $result) {
+            // Skip results of "SADD" operation, they'll be 1 or 0 depending on if set value already existed or not
+            if (is_numeric($result) && isset($tagSets[$id])) {
+                continue;
+            }
+            // setEx results
+            if (true !== $result && (!$result instanceof Status || $result !== Status::get('OK'))) {
+                $failed[] = $id;
+            }
+        }
+
+        return $failed;
+    }
+
+    /**
+     * This method overrides @see \Symfony\Component\Cache\Traits\RedisTrait::doFetch in order to use mget & marshaller.
+     *
+     * {@inheritdoc}
+     */
+    protected function doFetch(array $ids)
+    {
+        if (empty($ids)) {
+            return [];
+        }
+
+        // Using MGET for speed on RedisCluster as pipeline is not supported there
+        $values = $this->redis->mget($ids);
+        foreach ($values as $key => $v) {
+            // Not found items will have value as false, key will be same as on $ids
+            if ($v) {
+                yield $ids[$key] => $this->marshaller->unmarshall($v);
+            }
+        }
+    }
+
+    /**
+     * @param array $tags
+     *
+     * @return bool|void
+     */
+    public function invalidateTags(array $tags)
+    {
+        if (empty($tags)) {
+            return;
+        }
+
+        // Retrive and delete items in bulk of 10.000 at a time to not overflow buffers
+        $getId = $this->getId;
+        do {
+            $tagIdSets = $this->pipeline(function () use ($tags, $getId, $limit) {
+                foreach (array_unique($tags) as $tag) {
+                    // Requires Predis or PHP Redis 3.1.3+ (https://github.com/phpredis/phpredis/commit/d2e203a6)
+                    yield 'sPop' => [$getId(self::TAGS_PREFIX . $tag), self::BULK_INVALIDATION_POP_LIMIT];
+                }
+            });
+
+            $ids = array_unique(array_merge(...$tagIdSets));
```

Comment:
I'm getting `Cannot unpack Traversable with string keys` error here.

Reply:
Should be fixed now, 7e6be16 wrongly used array_merge directly on the generator, code is now back to a foreach.